arangod/Aql/OptimizerRules.cpp

```diff
@@ -1811,28 +1814,40 @@ void arangodb::aql::moveCalculationsDownRule(
       auto current = stack.back();
       stack.pop_back();
 
-      bool done = false;
+      auto const currentType = current->getType();
+      bool tryToPushIntoSubquery = false;
 
-      usedHere.clear();
-      current->getVariablesUsedHere(usedHere);
-      for (auto const& v : usedHere) {
-        if (v == variable) {
-          // the node we're looking at needs the variable we're setting.
-          // can't push further!
-          done = true;
-          break;
+      if (currentType == EN::SUBQUERY && !current->isVarUsedLater(variable)) {
+        current = ExecutionNode::castTo<SubqueryNode*>(current)->getSubquery();
+        while (current->hasDependency()) {
+          current = current->getFirstDependency();
+        }
+        tryToPushIntoSubquery = true;
+      } else {
+        usedHere.clear();
+        current->getVariablesUsedHere(usedHere);
+
+        bool done = false;
+        for (auto const& v : usedHere) {
+          if (v == variable) {
+            // the node we're looking at needs the variable we're setting.
+            // can't push further!
+            done = true;
+            break;
+          }
         }
-      }
 
-      if (done) {
-        // done with optimizing this calculation node
-        break;
+        if (done) {
+          // done with optimizing this calculation node
+          break;
+        }
       }
 
-      auto const currentType = current->getType();
-
-      if (currentType == EN::FILTER || currentType == EN::SORT ||
-          currentType == EN::LIMIT || currentType == EN::SUBQUERY) {
+      if (tryToPushIntoSubquery) {
+        lastNode = current;
+      } else if (currentType == EN::FILTER || currentType == EN::SORT ||
+                 currentType == EN::LIMIT || currentType == EN::SUBQUERY ||
+                 currentType == EN::SINGLETON) {
```

Comment:
Yes, it is necessary to include the SINGLETON node here now because the optimizer rule to push calculations down was extended to push calculations into a subquery if the calculation is **only** used by the subquery.
Consider this initial plan:
```
LET x = ...
LET sub1 = (
  FOR doc IN x
    ...
)
...
```
The previous implementation of the `moveCalculationsDown` rule did not allow to move `x` into the subquery, even if `x` was only used by the subquery and not afterwards.
With this PR, the rule can now move `x` into the subquery. 
In order to check if `x` can be moved into the subquery, we will make `current` point at the first node of the subquery, which is always of type SINGLETON (as all subqueries will start with a SINGLETON node).
If we don't add the SINGLETON node to the list of allowed nodes, no calculations will ever be moved into a subquery.